File: packages/shared/src/built-in-icon-index-builder.ts
import type { BuiltInIcon, BuiltInIconVariant } from "./media-resolver";
import type { MediaResolverConfig } from "./media-resolver-config";
import { BUILT_IN_ICON_PROVIDERS, type BuiltInIconProvider } from "./built-in-icons.ts";
import type { BuiltInIconProviderVersion, BuiltInIconSeedMetadata } from "./schemas/media";

export type BuiltInIconRegistryFetcher = (url: string, label: string) => Promise<unknown>;

type ProviderConfig = MediaResolverConfig["builtInProviders"][number];
export interface BuiltInIconProviderIndexSourceRef {
  provider: BuiltInIconProvider;
  cdnBase: string;
}

interface BuildContext {
  providers: Readonly<Record<BuiltInIconProvider, ProviderConfig>>;
  planSuffixWords: ReadonlySet<string>;
  fetchJson: BuiltInIconRegistryFetcher;
}

type JsonRecord = Record<string, unknown>;

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string").map((item) => item.trim()).filter(Boolean);
}

function isSafePathPart(value: string): boolean {
  return /^[a-z0-9][a-z0-9._-]*$/i.test(value);
}

export function normalizeBuiltInIconTerm(value: unknown): string {
  return String(value ?? "")
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/[^\p{L}\p{N}]+/gu, " ")
    .trim()
    .replace(/\s+/g, " ");
}

function compactTerm(value: unknown): string {
  return normalizeBuiltInIconTerm(value).replace(/\s+/g, "");
}

function uniqueTerms(values: readonly unknown[]): string[] {
  return [...new Set(values.map(normalizeBuiltInIconTerm).filter(Boolean))];
}

function exactKeysForIcon(icon: Pick<BuiltInIcon, "slug" | "title" | "aliases">): string[] {
  const canonical = uniqueTerms([icon.slug, icon.title, ...(icon.aliases ?? [])]);
  const compact = canonical.map(compactTerm).filter(Boolean);
  return [...new Set([...canonical, ...compact])];
}

function tokenKeysForIcon(
  icon: Pick<BuiltInIcon, "slug" | "title" | "aliases">,
  planSuffixWords: ReadonlySet<string>,
): string[] {
  const canonical = uniqueTerms([icon.slug, icon.title, ...(icon.aliases ?? [])]);
  return [...new Set(canonical
    .flatMap((term) => term.split(/\s+/))
    .filter((term) => term.length >= 3 && !planSuffixWords.has(term)))];
}

function iconRecord(input: Omit<BuiltInIcon, "terms" | "compactTerms" | "exactKeys" | "tokenKeys">, context: BuildContext): BuiltInIcon {
  return {
    ...input,
    terms: uniqueTerms([input.slug, input.title, ...(input.aliases ?? []), ...(input.categories ?? [])]),
    compactTerms: uniqueTerms([input.slug, input.title, ...(input.aliases ?? [])]).map(compactTerm).filter(Boolean),
    exactKeys: exactKeysForIcon(input),
    tokenKeys: tokenKeysForIcon(input, context.planSuffixWords),
  };
}

function optionalTheSvgMetadata(item: JsonRecord): Partial<Pick<BuiltInIcon, "hex" | "license" | "url" | "guidelines">> {
  const metadata: Partial<Pick<BuiltInIcon, "hex" | "license" | "url" | "guidelines">> = {};
  const hex = asString(item["hex"]);
  const license = asString(item["license"]);
  const url = asString(item["url"]);
  const guidelines = asString(item["guidelines"]);
  if (hex) metadata.hex = hex;
  if (license) metadata.license = license;
  if (url) metadata.url = url;
  if (guidelines) metadata.guidelines = guidelines;
  return metadata;
}

function parseTheSvgVariants(slug: string, value: unknown): BuiltInIconVariant[] {
  if (!isRecord(value)) return [];
  const variants: BuiltInIconVariant[] = [];
  for (const [variant, pathValue] of Object.entries(value)) {
    if (!isSafePathPart(variant)) continue;
    if (typeof pathValue !== "string") continue;
    const normalizedPath = pathValue.trim();
    if (!normalizedPath.endsWith(".svg")) continue;
    // 上游 registry 只提供路径事实，但仍必须锁在当前 slug 目录内，避免刷新入口把跨目录路径写入全局候选索引。
    if (!normalizedPath.startsWith(`/icons/${slug}/`)) continue;
    variants.push({ name: variant, path: `/public${normalizedPath}` });
  }
  return variants;
}

async function loadTheSvgIcons(context: BuildContext): Promise<BuiltInIcon[]> {
  const registry = await context.fetchJson(`${context.providers.thesvg.cdnBase}/src/data/icons.json`, "TheSVG registry");
  if (!Array.isArray(registry)) return [];
  const icons: BuiltInIcon[] = [];
  const seen = new Set<string>();
  for (const item of registry) {
    if (!isRecord(item)) continue;
    const slug = asString(item["slug"]);
    const title = asString(item["title"]);
    if (!slug || !title || !isSafePathPart(slug) || seen.has(slug)) continue;
    const variants = parseTheSvgVariants(slug, item["variants"]);
    if (variants.length === 0) continue;
    icons.push(iconRecord({
      provider: "thesvg",
      slug,
      title,
      aliases: asStringArray(item["aliases"]),
      categories: asStringArray(item["categories"]),
      variants,
      ...optionalTheSvgMetadata(item),
    }, context));
    seen.add(slug);
  }
  return icons;
}

function selfhstVariants(reference: string, item: JsonRecord): BuiltInIconVariant[] {
  const variants: BuiltInIconVariant[] = [];
  if (item["SVG"] === "Yes" || item["SVG"] === "Y") variants.push({ name: "default", path: `/svg/${reference}.svg` });
  if (item["Light"] === "Yes" || item["Light"] === "Y") variants.push({ name: "light", path: `/svg/${reference}-light.svg` });
  if (item["Dark"] === "Yes" || item["Dark"] === "Y") variants.push({ name: "dark", path: `/svg/${reference}-dark.svg` });
  return variants;
}

async function fetchJsonAny(context: BuildContext, urls: readonly string[], label: string): Promise<unknown> {
  const errors: string[] = [];
  for (const url of urls) {
    try {
      return await context.fetchJson(url, label);
    } catch (error) {
      errors.push(error instanceof Error ? error.message : String(error));
    }
  }
  throw new Error(`${label} failed: ${errors.join("; ")}`);
}

function githubRawBase(context: BuildContext, provider: BuiltInIconProvider): string {
  const config = context.providers[provider];
  const atIndex = config.cdnBase.lastIndexOf("@");
  const ref = atIndex >= 0 ? config.cdnBase.slice(atIndex + 1) : config.github.branch;
  return `https://raw.githubusercontent.com/${config.github.owner}/${config.github.repo}/${ref}`;
}

async function loadSelfhstIcons(context: BuildContext): Promise<BuiltInIcon[]> {
  const registry = await fetchJsonAny(context, [
    `${context.providers.selfhst.cdnBase}/index.json`,
    `${githubRawBase(context, "selfhst")}/index.json`,
  ], "selfh.st index");
  if (!Array.isArray(registry)) return [];
  const icons: BuiltInIcon[] = [];
  const seen = new Set<string>();
  for (const item of registry) {
    if (!isRecord(item)) continue;
    const slug = asString(item["Reference"]);
    const title = asString(item["Name"]);
    if (!slug || !title || !isSafePathPart(slug) || seen.has(slug)) continue;
    const variants = selfhstVariants(slug, item);
    if (variants.length === 0) continue;
    const tagTerms = asString(item["Tags"])?.split(",").map((value) => value.trim()).filter(Boolean) ?? [];
    const categories = [asString(item["Category"]), ...tagTerms].filter((value): value is string => Boolean(value));
    icons.push(iconRecord({
      provider: "selfhst",
      slug,
      title,
      aliases: [],
      categories,
      variants,
    }, context));
    seen.add(slug);
  }
  return icons;
}

function dashboardVariants(slug: string, item: JsonRecord, svgFiles: ReadonlySet<string>): BuiltInIconVariant[] {
  const variants: BuiltInIconVariant[] = [];
  if (svgFiles.has(`${slug}.svg`)) variants.push({ name: "default", path: `/svg/${slug}.svg` });
  const colors = item["colors"];
  if (isRecord(colors)) {
    for (const variantName of ["light", "dark"] as const) {
      const fileSlug = asString(colors[variantName]);
      if (!fileSlug || !isSafePathPart(fileSlug)) continue;
      if (svgFiles.has(`${fileSlug}.svg`)) variants.push({ name: variantName, path: `/svg/${fileSlug}.svg` });
    }
  }
  return variants;
}

function titleFromSlug(slug: string): string {
  return slug.split("-").filter(Boolean).map((part) => `${part.slice(0, 1).toUpperCase()}${part.slice(1)}`).join(" ");
}

async function loadDashboardIcons(context: BuildContext): Promise<BuiltInIcon[]> {
  const [metadata, tree] = await Promise.all([
    fetchJsonAny(context, [
      `${context.providers.dashboardIcons.cdnBase}/metadata.json`,
      `${githubRawBase(context, "dashboardIcons")}/metadata.json`,
    ], "Dashboard Icons metadata"),
    fetchJsonAny(context, [
      `${context.providers.dashboardIcons.cdnBase}/tree.json`,
      `${githubRawBase(context, "dashboardIcons")}/tree.json`,
    ], "Dashboard Icons tree"),
  ]);
  if (!isRecord(metadata) || !isRecord(tree) || !Array.isArray(tree["svg"])) return [];
  const svgFiles = new Set(tree["svg"].filter((item): item is string => typeof item === "string"));
  const icons: BuiltInIcon[] = [];
  for (const [slug, item] of Object.entries(metadata)) {
    if (!isSafePathPart(slug) || !isRecord(item)) continue;
    const variants = dashboardVariants(slug, item, svgFiles);
    if (variants.length === 0) continue;
    icons.push(iconRecord({
      provider: "dashboardIcons",
      slug,
      title: titleFromSlug(slug),
      aliases: asStringArray(item["aliases"]),
      categories: asStringArray(item["categories"]),
      variants,
    }, context));
  }
  return icons;
}

function createBuildContext(
  config: MediaResolverConfig,
  fetchJson: BuiltInIconRegistryFetcher,
  sourceRef?: BuiltInIconProviderIndexSourceRef,
): BuildContext {
  const providers = Object.fromEntries(config.builtInProviders.map((item) => [item.provider, item])) as Record<BuiltInIconProvider, ProviderConfig>;
  if (sourceRef) {
    providers[sourceRef.provider] = {
      ...providers[sourceRef.provider],
      cdnBase: sourceRef.cdnBase,
    };
  }
  return {
    providers,
    planSuffixWords: new Set(config.auto.planSuffixWords),
    fetchJson,
  };
}

export async function buildBuiltInIconIndex(
  config: MediaResolverConfig,
  fetchJson: BuiltInIconRegistryFetcher,
): Promise<BuiltInIcon[]> {
  const providerIndexes = await Promise.all(BUILT_IN_ICON_PROVIDERS.map(async (provider) => [
    provider,
    await buildBuiltInIconProviderIndex(config, provider, fetchJson),
  ] as const));
  return mergeBuiltInIconProviderIndexes(Object.fromEntries(providerIndexes) as Partial<Record<BuiltInIconProvider, readonly BuiltInIcon[]>>);
}

export async function buildBuiltInIconProviderIndex(
  config: MediaResolverConfig,
  provider: BuiltInIconProvider,
  fetchJson: BuiltInIconRegistryFetcher,
  sourceRef?: BuiltInIconProviderIndexSourceRef,
): Promise<BuiltInIcon[]> {
  const context = createBuildContext(config, fetchJson, sourceRef);
  const icons = provider === "thesvg"
    ? await loadTheSvgIcons(context)
    : provider === "selfhst"
      ? await loadSelfhstIcons(context)
      : await loadDashboardIcons(context);
  if (icons.length === 0) {
    throw new Error(`${provider} built-in icon index generation produced no icons`);
  }
  return icons;
}

export function mergeBuiltInIconProviderIndexes(
  providerIndexes: Partial<Record<BuiltInIconProvider, readonly BuiltInIcon[]>>,
): BuiltInIcon[] {
  const icons = BUILT_IN_ICON_PROVIDERS.flatMap((provider) => providerIndexes[provider] ?? []);
  if (icons.length === 0) {
    throw new Error("built-in icon index generation produced no icons");
  }
  return icons;
}

export function replaceBuiltInIconProviderIndex(
  icons: readonly BuiltInIcon[],
  provider: BuiltInIconProvider,
  providerIcons: readonly BuiltInIcon[],
): BuiltInIcon[] {
  return mergeBuiltInIconProviderIndexes({
    ...Object.fromEntries(BUILT_IN_ICON_PROVIDERS.map((item) => [
      item,
      item === provider ? providerIcons : icons.filter((icon) => icon.provider === item),
    ])) as Record<BuiltInIconProvider, readonly BuiltInIcon[]>,
  });
}

export function canonicalBuiltInIconIndexJson(icons: readonly BuiltInIcon[]): string {
  return `${JSON.stringify(icons)}\n`;
}

export function createBuiltInIconSeedMetadata(
  icons: readonly Pick<BuiltInIcon, "provider">[],
  hash: string,
  providers: Record<BuiltInIconProvider, BuiltInIconProviderVersion>,
): BuiltInIconSeedMetadata {
  return {
    hash,
    iconCount: icons.length,
    providerCounts: countBuiltInIconProviders(icons),
    providers,
  };
}

export function canonicalBuiltInIconSeedMetadataJson(metadata: BuiltInIconSeedMetadata): string {
  return `${JSON.stringify(metadata)}\n`;
}

export function countBuiltInIconProviders(icons: readonly Pick<BuiltInIcon, "provider">[]): Record<BuiltInIconProvider, number> {
  return icons.reduce<Record<BuiltInIconProvider, number>>((counts, icon) => ({
    ...counts,
    [icon.provider]: counts[icon.provider] + 1,
  }), {
    thesvg: 0,
    selfhst: 0,
    dashboardIcons: 0,
  });
}

File: packages/server/cmd/renewlet/built_in_icon_index_builder.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
)

const builtInIconRegistryFetchTimeout = 15 * time.Second
const builtInIconRegistryJSONLimitBytes = 16 * 1024 * 1024

var safeBuiltInIconPathPartRE = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]*$`)
var builtInIconIndexHTTPClient = &http.Client{Timeout: builtInIconRegistryFetchTimeout}

type builtInIconProviderSourceRef struct {
	Provider string
	CDNBase  string
}

func buildRemoteBuiltInIconIndex(ctx context.Context) ([]builtInIcon, error) {
	providerIndexes := map[string][]builtInIcon{}
	for _, provider := range []string{"thesvg", "selfhst", "dashboardIcons"} {
		icons, err := buildRemoteBuiltInIconProviderIndex(ctx, provider, nil)
		if err != nil {
			return nil, err
		}
		providerIndexes[provider] = icons
	}
	return mergeBuiltInIconProviderIndexes(providerIndexes)
}

func buildRemoteBuiltInIconProviderIndex(ctx context.Context, provider string, sourceRef *builtInIconProviderSourceRef) ([]builtInIcon, error) {
	var icons []builtInIcon
	var err error
	switch provider {
	case "thesvg":
		icons, err = loadTheSVGIcons(ctx, sourceRef)
	case "selfhst":
		icons, err = loadSelfhstIcons(ctx, sourceRef)
	case "dashboardIcons":
		icons, err = loadDashboardIcons(ctx, sourceRef)
	default:
		return nil, fmt.Errorf("unknown built-in icon provider: %s", provider)
	}
	if err != nil {
		return nil, err
	}
	if len(icons) == 0 {
		return nil, fmt.Errorf("%s built-in icon index generation produced no icons", provider)
	}
	return icons, nil
}

func mergeBuiltInIconProviderIndexes(providerIndexes map[string][]builtInIcon) ([]builtInIcon, error) {
	icons := []builtInIcon{}
	for _, provider := range []string{"thesvg", "selfhst", "dashboardIcons"} {
		icons = append(icons, providerIndexes[provider]...)
	}
	if len(icons) == 0 {
		return nil, errors.New("built-in icon index generation produced no icons")
	}
	return icons, nil
}

func replaceBuiltInIconProviderIndex(icons []builtInIcon, provider string, providerIcons []builtInIcon) ([]builtInIcon, error) {
	providerIndexes := map[string][]builtInIcon{}
	for _, icon := range icons {
		if icon.Provider == provider {
			continue
		}
		providerIndexes[icon.Provider] = append(providerIndexes[icon.Provider], icon)
	}
	providerIndexes[provider] = providerIcons
	return mergeBuiltInIconProviderIndexes(providerIndexes)
}

func builtInIconProviderSourceBase(provider string, sourceRef *builtInIconProviderSourceRef) string {
	if sourceRef != nil && sourceRef.Provider == provider && sourceRef.CDNBase != "" {
		return sourceRef.CDNBase
	}
	return mediaResolverBuiltInProviderBase(provider)
}

func builtInIconProviderGitHubRawBase(provider string, sourceRef *builtInIconProviderSourceRef) string {
	ref := "main"
	if sourceRef != nil && sourceRef.Provider == provider && sourceRef.CDNBase != "" {
		if at := strings.LastIndex(sourceRef.CDNBase, "@"); at >= 0 && at < len(sourceRef.CDNBase)-1 {
			ref = sourceRef.CDNBase[at+1:]
		}
	}
	for _, item := range mediaResolverCfg.BuiltInProviders {
		if item.Provider == provider && item.GitHub.Owner != "" && item.GitHub.Repo != "" {
			if sourceRef == nil && item.GitHub.Branch != "" {
				ref = item.GitHub.Branch
			}
			return "https://raw.githubusercontent.com/" + item.GitHub.Owner + "/" + item.GitHub.Repo + "/" + ref
		}
	}
	return ""
}

func fetchRegistryJSON(ctx context.Context, url string, label string, target any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	res, err := builtInIconIndexHTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("%s HTTP %d", label, res.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(res.Body, builtInIconRegistryJSONLimitBytes+1))
	if err != nil {
		return err
	}
	if len(data) > builtInIconRegistryJSONLimitBytes {
		return fmt.Errorf("%s response too large", label)
	}
	return json.Unmarshal(data, target)
}

func fetchRegistryJSONAny(ctx context.Context, urls []string, label string, target any) error {
	failures := []string{}
	for _, url := range urls {
		if err := fetchRegistryJSON(ctx, url, label, target); err != nil {
			failures = append(failures, err.Error())
			continue
		}
		return nil
	}
	return fmt.Errorf("%s failed: %s", label, strings.Join(failures, "; "))
}

func safeBuiltInIconPathPart(value string) bool {
	return safeBuiltInIconPathPartRE.MatchString(value)
}

func stringValue(value any) string {
	if text, ok := value.(string); ok {
		return strings.TrimSpace(text)
	}
	return ""
}

func stringSliceValue(value any) []string {
	items, ok := value.([]any)
	if !ok {
		return []string{}
	}
	out := []string{}
	for _, item := range items {
		if text := stringValue(item); text != "" {
			out = append(out, text)
		}
	}
	return out
}

func exactKeysForBuiltInIcon(slug string, title string, aliases []string) []string {
	canonical := normalizedTerms(append([]string{slug, title}, aliases...))
	compact := compactTerms(canonical)
	return uniqueStrings(append(canonical, compact...))
}

func iconRecord(input builtInIcon) builtInIcon {
	input.Terms = normalizedTerms(append(append([]string{input.Slug, input.Title}, input.Aliases...), input.Categories...))
	input.CompactTerms = compactTerms(append([]string{input.Slug, input.Title}, input.Aliases...))
	input.ExactKeys = exactKeysForBuiltInIcon(input.Slug, input.Title, input.Aliases)
	input.TokenKeys = exactTokenKeys(append([]string{input.Slug, input.Title}, input.Aliases...))
	return input
}

func loadTheSVGIcons(ctx context.Context, sourceRef *builtInIconProviderSourceRef) ([]builtInIcon, error) {
	var registry []map[string]any
	if err := fetchRegistryJSON(ctx, builtInIconProviderSourceBase("thesvg", sourceRef)+"/src/data/icons.json", "TheSVG registry", &registry); err != nil {
		return nil, err
	}
	icons := []builtInIcon{}
	seen := map[string]struct{}{}
	for _, item := range registry {
		slug := stringValue(item["slug"])
		title := stringValue(item["title"])
		if slug == "" || title == "" || !safeBuiltInIconPathPart(slug) {
			continue
		}
		if _, ok := seen[slug]; ok {
			continue
		}
		variants := theSVGVariants(slug, item["variants"])
		if len(variants) == 0 {
			continue
		}
		icons = append(icons, iconRecord(builtInIcon{
			Provider:   "thesvg",
			Slug:       slug,
			Title:      title,
			Aliases:    stringSliceValue(item["aliases"]),
			Categories: stringSliceValue(item["categories"]),
			Variants:   variants,
			Hex:        stringValue(item["hex"]),
			License:    stringValue(item["license"]),
			URL:        stringValue(item["url"]),
			Guidelines: stringValue(item["guidelines"]),
		}))
		seen[slug] = struct{}{}
	}
	return icons, nil
}

func theSVGVariants(slug string, value any) []builtInIconVariant {
	raw, ok := value.(map[string]any)
	if !ok {
		return []builtInIconVariant{}
	}
	variants := []builtInIconVariant{}
	for name, pathValue := range raw {
		path, ok := pathValue.(string)
		path = strings.TrimSpace(path)
		if !ok || !safeBuiltInIconPathPart(name) || !strings.HasSuffix(path, ".svg") {
			continue
		}
		// 上游 registry 只提供路径事实，但仍必须锁在当前 slug 目录内，避免刷新入口把跨目录路径写入全局候选索引。
		if !strings.HasPrefix(path, "/icons/"+slug+"/") {
			continue
		}
		variants = append(variants, builtInIconVariant{Name: name, Path: "/public" + path})
	}
	return variants
}

func loadSelfhstIcons(ctx context.Context, sourceRef *builtInIconProviderSourceRef) ([]builtInIcon, error) {
	var registry []map[string]any
	if err := fetchRegistryJSONAny(ctx, []string{
		builtInIconProviderSourceBase("selfhst", sourceRef) + "/index.json",
		builtInIconProviderGitHubRawBase("selfhst", sourceRef) + "/index.json",
	}, "selfh.st index", &registry); err != nil {
		return nil, err
	}
	icons := []builtInIcon{}
	seen := map[string]struct{}{}
	for _, item := range registry {
		slug := stringValue(item["Reference"])
		title := stringValue(item["Name"])
		if slug == "" || title == "" || !safeBuiltInIconPathPart(slug) {
			continue
		}
		if _, ok := seen[slug]; ok {
			continue
		}
		variants := selfhstIconVariants(slug, item)
		if len(variants) == 0 {
			continue
		}
		categories := []string{}
		if category := stringValue(item["Category"]); category != "" {
			categories = append(categories, category)
		}
		for _, tag := range strings.Split(stringValue(item["Tags"]), ",") {
			if trimmed := strings.TrimSpace(tag); trimmed != "" {
				categories = append(categories, trimmed)
			}
		}
		icons = append(icons, iconRecord(builtInIcon{
			Provider:   "selfhst",
			Slug:       slug,
			Title:      title,
			Aliases:    []string{},
			Categories: categories,
			Variants:   variants,
		}))
		seen[slug] = struct{}{}
	}
	return icons, nil
}

func selfhstIconVariants(reference string, item map[string]any) []builtInIconVariant {
	variants := []builtInIconVariant{}
	if item["SVG"] == "Yes" || item["SVG"] == "Y" {
		variants = append(variants, builtInIconVariant{Name: "default", Path: "/svg/" + reference + ".svg"})
	}
	if item["Light"] == "Yes" || item["Light"] == "Y" {
		variants = append(variants, builtInIconVariant{Name: "light", Path: "/svg/" + reference + "-light.svg"})
	}
	if item["Dark"] == "Yes" || item["Dark"] == "Y" {
		variants = append(variants, builtInIconVariant{Name: "dark", Path: "/svg/" + reference + "-dark.svg"})
	}
	return variants
}

func loadDashboardIcons(ctx context.Context, sourceRef *builtInIconProviderSourceRef) ([]builtInIcon, error) {
	var metadata map[string]map[string]any
	if err := fetchRegistryJSONAny(ctx, []string{
		builtInIconProviderSourceBase("dashboardIcons", sourceRef) + "/metadata.json",
		builtInIconProviderGitHubRawBase("dashboardIcons", sourceRef) + "/metadata.json",
	}, "Dashboard Icons metadata", &metadata); err != nil {
		return nil, err
	}
	var tree struct {
		SVG []string `json:"svg"`
	}
	if err := fetchRegistryJSONAny(ctx, []string{
		builtInIconProviderSourceBase("dashboardIcons", sourceRef) + "/tree.json",
		builtInIconProviderGitHubRawBase("dashboardIcons", sourceRef) + "/tree.json",
	}, "Dashboard Icons tree", &tree); err != nil {
		return nil, err
	}
	svgFiles := map[string]struct{}{}
	for _, file := range tree.SVG {
		svgFiles[file] = struct{}{}
	}
	icons := []builtInIcon{}
	for slug, item := range metadata {
		if !safeBuiltInIconPathPart(slug) {
			continue
		}
		variants := dashboardIconVariants(slug, item, svgFiles)
		if len(variants) == 0 {
			continue
		}
		icons = append(icons, iconRecord(builtInIcon{
			Provider:   "dashboardIcons",
			Slug:       slug,
			Title:      titleFromBuiltInIconSlug(slug),
			Aliases:    stringSliceValue(item["aliases"]),
			Categories: stringSliceValue(item["categories"]),
			Variants:   variants,
		}))
	}
	return icons, nil
}

func dashboardIconVariants(slug string, item map[string]any, svgFiles map[string]struct{}) []builtInIconVariant {
	variants := []builtInIconVariant{}
	if _, ok := svgFiles[slug+".svg"]; ok {
		variants = append(variants, builtInIconVariant{Name: "default", Path: "/svg/" + slug + ".svg"})
	}
	colors, ok := item["colors"].(map[string]any)
	if !ok {
		return variants
	}
	for _, variantName := range []string{"light", "dark"} {
		fileSlug := stringValue(colors[variantName])
		if fileSlug == "" || !safeBuiltInIconPathPart(fileSlug) {
			continue
		}
		if _, ok := svgFiles[fileSlug+".svg"]; ok {
			variants = append(variants, builtInIconVariant{Name: variantName, Path: "/svg/" + fileSlug + ".svg"})
		}
	}
	return variants
}

func titleFromBuiltInIconSlug(slug string) string {
	parts := strings.Split(slug, "-")
	for index, part := range parts {
		if part == "" {
			continue
		}
		parts[index] = strings.ToUpper(part[:1]) + part[1:]
	}
	return strings.Join(parts, " ")
}
